### PR TITLE
Create the SITE_MIGRATION_FALLBACK_CREDENTIALS step

### PR DESCRIPTION
**Pull request by andregardi**
## Proposed Changes
image

As part of application passwords integration on migrations, we are introducing the new the SITE_MIGRATION_FALLBACK_CREDENTIALS step.
It is shares some behavior with the current credential steps, but does not prompt the user for url and does no offer the back file as alternative method.

We present this step to the user when:
- The `automated-migration/application-password` feature flag is enabled.
- The migrated site does not have application password available.

We are making a request to `/automated-migration/application-passwords/authorization-url`, that will be used to check if the site has application passwords available. 
- If it is disabled, we are directing the user to the `SITE_MIGRATION_FALLBACK_CREDENTIALS` step.
- If it is available, we are directing the user to the `SITE_MIGRATION_APPLICATION_PASSWORDS_APPROVAL` step, and adding the `authorizationUrl` as a parameter to be used on next step.

## Testing Instructions

The `automated-migration/application-password` feature flag is already on localhost environment.

### Testing the `authorization-url` endpoint call and logic.
We are using the [Disable Application Passwords](https://wordpress.org/plugins/disable-application-passwords/) plugin to test the scenario in which applications password is not available.

Start a **Do it for me** migration until you reach the `/site-migration-credentials` step.

Click on **Continue** button.
- If the plugin is disabled, you should be directed to the `SITE_MIGRATION_APPLICATION_PASSWORDS_APPROVAL` step. This is still a placeholder step, don't expect any functionality to be implemented. Check that the ``authorizationUrl` parameter is added to the url.
- If the plugin is enabled, you should be directed to the `SITE_MIGRATION_FALLBACK_CREDENTIALS` step.

### Testing SITE_MIGRATION_FALLBACK_CREDENTIALS step

Try to input a wrong username and password.
Click on **Continue** button.
You should get the following error.
image

The button label should change to Continue anyway.
Clicking on it should create a migration ticket on Zendesk.

Close the Zendesk ticket and hit the back button.
This time use the right username and password.
It should also create a Zendesk ticket. This time with the  **quickforget** link to the credentials.

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

**Comment by andres-blanco** [comment]
Hi @andregardi, Will you update the name of the step as part of this PR?

**Comment by sixhours** [comment]
This is working for me!

File: client/landing/stepper/declarative-flow/internals/steps-repository/site-migration-fallback-credentials/components/credentials-form.tsx [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+			'We ran into a problem submitting your details. Please try again shortly.'
+		);
+	}
+
+	const getContinueButtonText = () => {
+		if ( isBusy && ! canBypassVerification ) {
+			return translate( 'Verifying credentials' );
+		}
+		if ( canBypassVerification ) {
+			return translate( 'Continue anyways' );
```
Comment (sixhours) [inline client/landing/stepper/declarative-flow/internals/steps-repository/site-migration-fallback-credentials/components/credentials-form.tsx]: I believe this should read "Continue anyway"

File: client/landing/stepper/declarative-flow/internals/steps-repository/site-migration-fallback-credentials/hooks/use-form-error-mapping.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+						type: 'special',
+						message: translate(
+							'We could not verify your credentials. Can you double check your account information and try again?'
+						),
+					},
+					...credentialsErrorMessage,
+				};
+			}
+
+			if ( code !== 'rest_invalid_param' || ! data?.params ) {
```
Comment (sixhours) [inline client/landing/stepper/declarative-flow/internals/steps-repository/site-migration-fallback-credentials/hooks/use-form-error-mapping.ts]: `code !== 'rest_invalid_param'` should be reversed to use Yoda conditionals.

File: client/landing/stepper/declarative-flow/internals/steps-repository/site-migration-fallback-credentials/hooks/use-form-error-mapping.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+			if ( code === 'rest_missing_callback_param' || ! code ) {
+				return {
+					root: {
+						type: 'manual',
+						message: translate( 'An error occurred while saving credentials.' ),
+					},
+				};
+			}
+
+			if ( code === 'automated_migration_tools_login_and_get_cookies_test_failed' ) {
```
Comment (sixhours) [inline client/landing/stepper/declarative-flow/internals/steps-repository/site-migration-fallback-credentials/hooks/use-form-error-mapping.ts]: `code === 'automated_migration_tools_login_and_get_cookies_test_failed'` should be reversed to use Yoda conditionals.

File: client/landing/stepper/declarative-flow/internals/steps-repository/site-migration-fallback-credentials/hooks/use-form-error-mapping.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+			);
+		},
+		[ fieldMapping, translate ]
+	);
+
+	const handleServerError = useCallback(
+		( error: ApiError ) => {
+			const { code, message, data } = error;
+
+			if ( code === 'rest_missing_callback_param' || ! code ) {
```
Comment (sixhours) [inline client/landing/stepper/declarative-flow/internals/steps-repository/site-migration-fallback-credentials/hooks/use-form-error-mapping.ts]: `code === 'rest_missing_callback_param'` should be reversed to use Yoda conditionals.

Comment (andregardi) [reply client/landing/stepper/declarative-flow/internals/steps-repository/site-migration-fallback-credentials/hooks/use-form-error-mapping.ts]: Thank you! Just reversed it to Yoda.

File: client/landing/stepper/declarative-flow/internals/steps-repository/site-migration-fallback-credentials/hooks/use-form-error-mapping.ts (hunk 3 above)
Comment (andregardi) [reply client/landing/stepper/declarative-flow/internals/steps-repository/site-migration-fallback-credentials/hooks/use-form-error-mapping.ts]: Thank you! Just reversed it to Yoda.

File: client/landing/stepper/declarative-flow/internals/steps-repository/site-migration-fallback-credentials/hooks/use-form-error-mapping.ts (hunk 2 above)
Comment (andregardi) [reply client/landing/stepper/declarative-flow/internals/steps-repository/site-migration-fallback-credentials/hooks/use-form-error-mapping.ts]: Thank you! Just reversed it to Yoda.

File: client/landing/stepper/declarative-flow/internals/steps-repository/site-migration-fallback-credentials/components/credentials-form.tsx (hunk 1 above)
Comment (andregardi) [reply client/landing/stepper/declarative-flow/internals/steps-repository/site-migration-fallback-credentials/components/credentials-form.tsx]: Sure, just updated it!

File: client/landing/stepper/declarative-flow/internals/steps-repository/site-migration-fallback-credentials/hooks/use-credentials-form.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
+import { useEffect, useState } from 'react';
+import { useForm } from 'react-hook-form';
+import { useQuery } from 'calypso/landing/stepper/hooks/use-query';
+import { useSiteSlugParam } from 'calypso/landing/stepper/hooks/use-site-slug-param';
+import { recordTracksEvent } from 'calypso/lib/analytics/tracks';
+import { CredentialsFormData } from '../../site-migration-credentials/types';
+import { useFormErrorMapping } from './use-form-error-mapping';
+import { useRequestAutomatedMigration } from './use-request-automated-migration';
+
+export const useCredentialsForm = ( onSubmit: ( from?: string ) => void ) => {
```
Comment (valterlorran) [inline client/landing/stepper/declarative-flow/internals/steps-repository/site-migration-fallback-credentials/hooks/use-credentials-form.ts]: Nit: I think this should be called useFallbackCredentialsForm to reflect that this is related to the fallback.

File: client/landing/stepper/declarative-flow/internals/steps-repository/site-migration-fallback-credentials/test/index.tsx [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+		post: jest.fn(),
+	},
+} ) );
+
+jest.mock( 'wpcom-proxy-request', () => jest.fn() );
+jest.mock( 'calypso/landing/stepper/hooks/use-site-slug-param' );
+
+( useSiteSlugParam as jest.Mock ).mockImplementation( () => 'site-url.wordpress.com' );
+
+// jest.mock( 'calypso/landing/stepper/hooks/use-query', () => ( {
```
Comment (valterlorran) [inline client/landing/stepper/declarative-flow/internals/steps-repository/site-migration-fallback-credentials/test/index.tsx]: can we remove this?

**Review by valterlorran — COMMENTED** [review]
Overall is looking good for me, I just a left a few nit comments.

File: client/landing/stepper/declarative-flow/internals/steps-repository/site-migration-fallback-credentials/test/index.tsx (hunk 6 above)
Comment (andregardi) [reply client/landing/stepper/declarative-flow/internals/steps-repository/site-migration-fallback-credentials/test/index.tsx]: Thank you for catching that! Removed now.

File: client/landing/stepper/declarative-flow/internals/steps-repository/site-migration-fallback-credentials/hooks/use-credentials-form.ts (hunk 5 above)
Comment (andregardi) [reply client/landing/stepper/declarative-flow/internals/steps-repository/site-migration-fallback-credentials/hooks/use-credentials-form.ts]: That is a good call and is quick to address. Just renamed the hook.

**Review by andres-blanco — APPROVED** [review]

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/17044681

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @andregardi for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/17044681).

**Pull request merged**